File: parser/tests/workflow-replay.test.ts
import { describe, it, expect } from 'vitest';
import { WorkflowReplayEngine } from '../src/workflow-replay';

describe('v58.0.0 WorkflowReplayEngine — Temporal Workflow Replay & Time-Travel Debugging', () => {
  it('captures an execution trace step-by-step', () => {
    const engine = new WorkflowReplayEngine();
    const trace = engine.startTrace('workflow-build-1');

    const step1 = engine.captureStep(trace.traceId, 'compile', 'agent-compiler', { file: 'index.ts' }, 'Success');
    const step2 = engine.captureStep(trace.traceId, 'test', 'agent-tester', { suite: 'unit' }, 'Passed 10 tests');
    engine.completeTrace(trace.traceId);

    expect(step1?.stepIndex).toBe(0);
    expect(step2?.stepIndex).toBe(1);

    const fetched = engine.getTrace(trace.traceId);
    expect(fetched?.steps.length).toBe(2);
    expect(fetched?.status).toBe('COMPLETED');
  });

  it('supports time-travel seek and step navigation', () => {
    const engine = new WorkflowReplayEngine();
    const trace = engine.startTrace('workflow-seek');

    engine.captureStep(trace.traceId, 'step-0', 'agent-1');
    engine.captureStep(trace.traceId, 'step-1', 'agent-1');
    engine.captureStep(trace.traceId, 'step-2', 'agent-1');
    engine.completeTrace(trace.traceId);

    const sought = engine.seekToStep(trace.traceId, 1);
    expect(sought?.action).toBe('step-1');

    const forward = engine.stepForward(trace.traceId);
    expect(forward?.action).toBe('step-2');

    const backward = engine.stepBackward(trace.traceId);
    expect(backward?.action).toBe('step-1');
  });

  it('compares two execution traces for state divergence', () => {
    const engine = new WorkflowReplayEngine();

    const traceA = engine.startTrace('run-a');
    engine.captureStep(traceA.traceId, 'build', 'agent-1');
    engine.captureStep(traceA.traceId, 'test', 'agent-2');
    engine.completeTrace(traceA.traceId);

    const traceB = engine.startTrace('run-b');
    engine.captureStep(traceB.traceId, 'build', 'agent-1');
    engine.captureStep(traceB.traceId, 'deploy', 'agent-2'); // Divergent action
    engine.completeTrace(traceB.traceId);

    const diffs = engine.compareTraces(traceA.traceId, traceB.traceId);
    expect(diffs.length).toBe(2);
    expect(diffs[0].hasStateDivergence).toBe(false);
    expect(diffs[1].hasStateDivergence).toBe(true);
    expect(diffs[1].baseAction).toBe('test');
    expect(diffs[1].compareAction).toBe('deploy');
  });
});

File: sdk/go/workflow_replay_test.go
package alpgo

import (
	"testing"
)

func TestWorkflowReplayEngine(t *testing.T) {
	engine := NewWorkflowReplayEngine()

	// Start trace
	trace := engine.StartTrace("wf-daily-etl")
	if trace.Status != ReplayCapturing {
		t.Fatalf("expected status %s, got %s", ReplayCapturing, trace.Status)
	}

	// Capture steps
	s1 := engine.CaptureStep(trace.TraceID, "extract_data", "agent-etl", map[string]interface{}{"rows": 100}, "ok")
	if s1 == nil || s1.StepIndex != 0 {
		t.Fatalf("expected step index 0, got %v", s1)
	}

	s2 := engine.CaptureStep(trace.TraceID, "transform_data", "agent-etl", map[string]interface{}{"cleaned": true}, "ok")
	if s2 == nil || s2.StepIndex != 1 {
		t.Fatalf("expected step index 1, got %v", s2)
	}

	// Complete trace
	ok := engine.CompleteTrace(trace.TraceID)
	if !ok || trace.Status != ReplayCompleted {
		t.Fatalf("expected completed status, got %s", trace.Status)
	}

	// Time travel seek
	seekStep := engine.SeekToStep(trace.TraceID, 0)
	if seekStep == nil || seekStep.Action != "extract_data" {
		t.Fatalf("expected extract_data, got %v", seekStep)
	}

	// Step forward
	fwdStep := engine.StepForward(trace.TraceID)
	if fwdStep == nil || fwdStep.Action != "transform_data" {
		t.Fatalf("expected transform_data, got %v", fwdStep)
	}

	// Step backward
	backStep := engine.StepBackward(trace.TraceID)
	if backStep == nil || backStep.Action != "extract_data" {
		t.Fatalf("expected extract_data on backward, got %v", backStep)
	}

	// Compare traces
	trace2 := engine.StartTrace("wf-daily-etl-v2")
	engine.CaptureStep(trace2.TraceID, "extract_data", "agent-etl", nil, "ok")
	engine.CaptureStep(trace2.TraceID, "transform_parquet_v2", "agent-etl", nil, "ok")
	engine.CompleteTrace(trace2.TraceID)

	diffs := engine.CompareTraces(trace.TraceID, trace2.TraceID)
	if len(diffs) != 2 {
		t.Fatalf("expected 2 diff steps, got %d", len(diffs))
	}
	if diffs[0].HasStateDivergence {
		t.Errorf("step 0 should match")
	}
	if !diffs[1].HasStateDivergence {
		t.Errorf("step 1 should diverge")
	}
}
